Mathlib/NumberTheory/ModularForms/Basic.lean

```diff
@@ -397,3 +399,125 @@ instance (priority := 99) [CuspFormClass F Γ k] : ModularFormClass F Γ k where
+  apply heq_of_cast_eq (type_eq Γ (add_zero k).symm).symm
+  funext
+  rw [cast_eq_mcast, mcast]
+  simp only [mul_coe, one_coe_eq_one, mul_one]
+  ext1
+  rfl
+  simp only [add_zero]
+
+theorem hEq_mul_assoc {a b c : ℤ} {Γ : Subgroup SL(2, ℤ)} (f : ModularForm Γ a)
+    (g : ModularForm Γ b) (h : ModularForm Γ c) : HEq ((f.mul g).mul h) (f.mul (g.mul h)) := by
+  apply heq_of_cast_eq (type_eq Γ (add_assoc a b c))
+  rw [cast_eq_mcast, mcast]
+  ext1
+  simp only [mul_coe, Pi.mul_apply, ← mul_assoc]
+  rfl
+  apply add_assoc
+
+theorem hEq_mul_comm {a b : ℤ} {Γ : Subgroup SL(2, ℤ)} (f : ModularForm Γ a) (g : ModularForm Γ b) :
+    HEq (f.mul g) (g.mul f) := by
+  apply heq_of_cast_eq (type_eq Γ (add_comm a b))
+  rw [cast_eq_mcast, mcast]
+  ext1
+  simp only [mul_coe, Pi.mul_apply, mul_comm]
+  rfl
+  apply add_comm
+
+variable (ι : Type _) (A : ι → Type _) [Zero ι] [GradedMonoid.GOne A]
+
+@[simp]
+theorem one_fst : (1 : GradedMonoid A).fst = 0 :=
+  rfl
+
+@[simp]
+theorem one_snd : (1 : GradedMonoid A).snd = 1 :=
+  rfl
+
+instance (Γ : Subgroup SL(2, ℤ)) : GradedMonoid.GOne (ModularForm Γ) where
+  one := 1
+
+instance (Γ : Subgroup SL(2, ℤ)) : GradedMonoid.GMul (ModularForm Γ) where
+  mul f g := f.mul g
+
+instance gradedModRing (Γ : Subgroup SL(2, ℤ)) : DirectSum.GCommRing (ModularForm Γ) where
+  mul f g := f.mul g
+  one := 1
+  one_mul f := by
+    rw [GradedMonoid.GOne.toOne, GradedMonoid.GMul.toMul]
+    apply Sigma.ext
+    · simp only [(· * ·), one_fst, zero_add]
+    · simp only [(· * ·), one_fst, one_snd, Submodule.coe_mk, one_mul, hEq_one_mul]
+  mul_one f := by
+    rw [GradedMonoid.GOne.toOne, GradedMonoid.GMul.toMul]
+    apply Sigma.ext
+    · simp only [(· * ·), one_fst, add_zero]
+    · simp only [(· * ·), one_fst, one_snd, Submodule.coe_mk, mul_one, hEq_mul_one]
+  mul_assoc f g h := by
+    rw [GradedMonoid.GMul.toMul]
+    apply Sigma.ext
+    · apply add_assoc
+    · simp only [(· * ·), Submodule.coe_mk, hEq_mul_assoc]
+  mul_zero {i j} f := by ext1; simp
+  zero_mul {i j} f := by ext1; simp
+  mul_add {i j} f g h := by
+    ext1
+    simp only [Pi.mul_apply, mul_add, mul_coe, add_apply]
+  add_mul {i j} f g h := by
+    ext1
+    simp only [add_mul, mul_coe, Pi.mul_apply, add_apply]
+  mul_comm f g := by
+    rw [GradedMonoid.GMul.toMul]
+    apply Sigma.ext
+    · apply add_comm
+    · apply hEq_mul_comm
+  gnpow_zero' := by
+    intro f
+    apply Sigma.ext <;> rw [GradedMonoid.GMonoid.gnpowRec_zero]
+  gnpow_succ' := by
+    intro n f
+    rw [GradedMonoid.GMul.toMul]
+    apply Sigma.ext <;> rw [GradedMonoid.GMonoid.gnpowRec_succ]
+  natCast n := n • (1 : ModularForm Γ 0)
+  natCast_zero := by simp only [zero_smul]
+  natCast_succ n := by simp only [add_smul, one_nsmul, add_right_inj]
+  intCast n := n • (1 : ModularForm Γ 0)
+  intCast_ofNat := by simp only [coe_nat_zsmul, forall_const]
+  intCast_negSucc_ofNat n := by simp only [Int.negSucc_coe]; apply _root_.neg_smul
```

Comment:
Can you create a separate `NatCast (ModularForm Γ 0)` instance first? Ideally you'd build it up from `SlashInvariantForm` to have a nice defeq.

Reply:
Sure, I did this for `IntCast` instead, since I guess thats the better instance to have?

Reply:
I think we need both for defeqs>

Reply:
Oh so you mean make each of them an separate instance and I guess add some lemma saying they agree?

Reply:
There's no need to add the lemma that says they agree, that'd handled automatically by existing lemmas in mathlib.

Reply:
alrighty, done! I didn't quite build it up from Slashinvariantform, but added a lemma saying they agree. But maybe thats a bad idea? I'm not yet sure what the problems would be of doing it this way.